### ISSUE TITLE
Request: Non-VPN mode

### ISSUE THREAD

**Issue by lars18th** [sshuttle/sshuttle#395]
Hi,

I readed the documentation multiple times. In special this part: https://sshuttle.readthedocs.io/en/stable/how-it-works.html
where it comments: _"sshuttle assembles the TCP stream locally, multiplexes it statefully over an ssh session, and disassembles it back into packets at the other end. So it never ends up doing TCP-over-TCP. It’s just data-over-TCP, which is safe"_.

So, compared to a simple "ssh -D 1080" command the `shuttle` improves the performance using some techniques to multiplexing different flows and assign to it different queues.

However, in the README page this project comments:
_"The remote network has no VPN, or only stupid/complex VPN protocols (IPsec, PPTP, etc). Or maybe you are the admin and you just got frustrated with the awful state of VPN tools."_
And I'm one of these users that _don't want to use VPNs_ to connect to some remote networks. I prefer to use simple port forwadings with tunnels.

However the performance of `sshuttle` is better than the simple "ssh -D 1080". So assuming this... Why not implement an standalone (aka non-VPN) mode? Please, note that this request is related only to the client part. The idea is to open a simple SOCKS4 server in the client machine.

Please consider this idea.
Regards.

**Comment by brianmay** [comment]
I actually find it hard to believe that `sshuttle` - written in python - and uses ssh - could be faster then `ssh -D` which is written in C.

Plus there are a lot of tickets here that the performance of sshuttle is terrible.

However if you want to write a SOCKS4 server for sshuttle, then please don't let me stop you :-)

**Comment by lars18th** [comment]
> I actually find it hard to believe that `sshuttle` - written in python - and uses ssh - could be faster then `ssh -D` which is written in C.

:laughing: Seriously! `shuttle` works in an upper layer than the transport, and it uses SSH as the transport. So it's not a problem to be developed in python. In fact, the _valuable_ characteristic of it is based on the use of _queues_ or _priorities_ to schedule the packets overt the SSH tunnel. And this is a functionality *NOT* implemented inside the ssh tool. So for this reason it can be good to have a SOCKS4 server inside the `shuttle`.

And regarding the performance of shuttle, that's the problem of _every_ VPN over TCP. Using the tunnelling approach is preferable than network (L2 or L3) encapsulation over a TCP channel. So, don't think that `sshuttle` is bad.

In any case. I hope someone (I don't have the time for it) will want to develop the non-VPN mode (aka SOCKS4 server) in `shuttle`.
Regards.

**Comment by shawnz** [comment]
The documentation seems to advertise that one of sshuttle's goals is to be more efficient than native SSH tunneling. I don't know if that is true or not, but it is claimed to be a use case of this application:

> You hate openssh’s port forwarding because it’s randomly slow and/or stupid. ... You can’t use openssh’s PermitTunnel feature because ... it does TCP-over-TCP, which has terrible performance

However that use case is not important for me personally. Such a feature would be helpful for me because of sshuttle's ability to bypass "AllowTcpForwarding" being disabled on the SSH server. Please consider it! Sshuttle has many potential use cases that don't really necessitate the client requiring admin rights or modifying the network configuration, which are necessary with the existing modes.

**Comment by ge9** [comment]
Since sshuttle supports UDP proxy in TPROXY mode, I guess SOCKS5 UDP associate can be supported in sshuttle client side.
SOCKS5 with UDP associate can be used as a transparent proxy in Linux (with TPROXY and https://github.com/semigodking/redsocks or https://github.com/heiher/hev-socks5-tproxy) and Windows (with https://github.com/wiresock/proxifyre , which support split proxying by application name).
This would surely be a great feature.

**Comment by brianmay** [comment]
At one stage, I was looking at doing a rewrite of sshuttle in rust, that uses TPROXY and talks to SOCKS5 instead of ssh + copy Python code.

I haven't looked at it for a while. Too many projects and not enough time. But it is available on github. https://github.com/sshuttle/sshuttle_rust

**Comment by ge9** [comment]
Yes, I have checked it but it's backed by ssh -D ? I hope something supporting UDP.

As an experiment, I ran Dante SOCKS5 server on the remote side and used https://github.com/mullvad/udp-over-tcp (written in Rust) to expose remote UDP port on the local machine.  I used Proxifyre as transparent proxy client and this basically worked as SOCKS5 proxy (with UDP associate support) in the local machine.

However, udp2tcp (in udp-over-tcp) only communicate with UDP client that sent packet to it first, so we need some more dynamic port assignment to make this system work continuously.

Since udp2tcp is written in Rust, I think it's good idea to porting sshuttle to Rust.

**Comment by brianmay** [comment]
It will work with any socks server, not just `ssh -D`.

It looks like I did have problems with UDP support, see https://github.com/sshuttle/sshuttle_rust?tab=readme-ov-file#udpdns-notes

But would be willing to merge changes that fix this (if this is possible).
